Mathlib/RingTheory/GradedAlgebra/HomogeneousRelation.lean

```diff
@@ -0,0 +1,322 @@
+/-
+Copyright (c) 2025 Jingting Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yiming Fu, Zhenyan Fu, Raphael Douglas Giles, Jiedong Jiang, Jingting Wang
+-/
+import Mathlib.Algebra.RingQuot
+import Mathlib.RingTheory.GradedAlgebra.Basic
+
+/-!
+# Homogeneous Relation
+
+In this file, we define the property of an ideal being homogeneous.
+
+## Main Definitions
+
+* `IsHomogeneousRelation`: a relation `r : A → A → Prop` is homogeneous with respect to the grading
+  `𝒜` if `r x y` implies that for every index `i`, the projection of `x` and `y` onto the `i`-th
+  grading piece satisfies the equivalence relation generated by `r`.
+
+## Main Results
+
+* We proved that taking `RingQuot` of a graded ring (or graded algebra) by a homogeneous relation
+  gives a graded strucutre on the quotient ring.
+-/
+
+variable {ι : Type*} [DecidableEq ι] [AddMonoid ι]
+variable {A : Type*} [Semiring A]
+
+/-- a relation `r : A → A → Prop` is homogeneous with respect to the grading `𝒜` if `r x y` implies
+  that for every index `i`, the projection of `x` and `y` onto the `i`-th grading piece satisfies
+  the equivalence relation generated by `r`. -/
+class IsHomogeneousRelation {σ : Type*} [SetLike σ A] [AddSubmonoidClass σ A] (𝒜 : ι → σ)
+  [GradedRing 𝒜] (r : A → A → Prop) : Prop where
+  is_homogeneous' : ∀ {x y : A}, r x y →
+  ∀ i : ι, (Relation.EqvGen r) ((GradedRing.proj 𝒜 i) x) ((GradedRing.proj 𝒜 i) y)
+
+namespace HomogeneousRelation
+
+section RingCon
+
+variable {σ : Type*} [SetLike σ A] [AddSubmonoidClass σ A]
+variable (𝒜 : ι → σ) [GradedRing 𝒜] (rel : A → A → Prop)
+
+open Relation GradedRing
+
+lemma eqvGen_ringQuot_of_eqvGen {a b : A} (h : EqvGen rel a b) :
+    EqvGen (RingQuot.Rel rel) a b := Relation.EqvGen.mono (fun _ _ h' ↦ RingQuot.Rel.of h') h
+
+lemma eqvGen_ringQuot_add_right {a b c : A} (h : EqvGen (RingQuot.Rel rel) a b) :
+    EqvGen (RingQuot.Rel rel) (a + c) (b + c) := by
+  induction h with
+  | rel x y hxy =>
+  apply EqvGen.rel
+  exact RingQuot.Rel.add_left hxy
```

Comment:
nit: these should be indented more so that they start at the same level as `rel x y hxy =>`